src/unix/aio.c

```diff
@@ -0,0 +1,352 @@
+            sizeof(*aio));
+    if (aio == INVALID_ADDRESS) {
+        return 0;
+    }
+    list_push_back(&p->aio, &aio->elem);
+    return aio;
+}
+
+static void aio_dealloc(process p, struct aio *aio)
+{
+    list_delete(&aio->elem);
+    deallocate(heap_general(get_kernel_heaps()), aio, sizeof(*aio));
+}
+
+static unsigned int aio_get_id(process p, struct aio *aio)
+{
+    unsigned int index = 0;
+    list_foreach(&p->aio, elem) {
+        if (elem == &aio->elem) {
+            return index;
+        }
+        index++;
+    }
+    return -1U;
+}
+
+static struct aio *aio_from_ring(process p, aio_ring ring)
+{
+    unsigned int index = 0;
+    list_foreach(&p->aio, elem) {
+        if (index == ring->id) {
+            return (struct aio *) elem;
+        }
+        index++;
+    }
+    return 0;
+}
+
+sysreturn io_setup(unsigned int nr_events, aio_context_t *ctx_idp)
+{
+    if (!ctx_idp) {
+        return -EFAULT;
+    }
+    if (nr_events == 0) {
+        return -EINVAL;
+    }
+
+    /* Allocate AIO ring structure and add it to process memory map.*/
+    kernel_heaps kh = get_kernel_heaps();
+    heap h = (heap) current->p->virtual_page;
+    aio_ring ctx;
+    nr_events += 1; /* needed because of head/tail management in ring buffer */
+    u64 alloc_size = pad(sizeof(*ctx) + nr_events * sizeof(struct io_event),
+            PAGESIZE);
+    ctx = (aio_ring) allocate_u64(h, alloc_size);
+    if (ctx == INVALID_ADDRESS) {
+        return -ENOMEM;
+    }
+    u64 phys = allocate_u64((heap) heap_physical(kh), alloc_size);
+    if (phys == INVALID_PHYSICAL) {
+        deallocate(h, ctx, alloc_size);
+        return -ENOMEM;
+    }
+    map(u64_from_pointer(ctx), phys, alloc_size,
+            PAGE_WRITABLE | PAGE_NO_EXEC | PAGE_USER, heap_pages(kh));
+
+    struct aio *aio = aio_alloc(current->p);
+    assert(aio);
+    aio->h = h;
+    aio->kh = kh;
+    aio->ring = ctx;
+    spin_lock_init(&aio->lock);
+    aio->bq = 0;
+    aio->nr = nr_events;
+    aio->ongoing_ops = 0;
+
+    ctx->id = aio_get_id(current->p, aio);
+    ctx->nr = nr_events;
+    ctx->head = ctx->tail = 0;
+    ctx->magic = AIO_RING_MAGIC;
+    ctx->compat_features = 1;   /* same as Linux kernel */
+    ctx->incompat_features = 0; /* same as Linux kernel */
+    ctx->header_length = sizeof(*ctx);
+    *ctx_idp = ctx;
+    return 0;
+}
+
+closure_function(2, 2, void, aio_eventfd_complete,
+                 heap, h, u64 *, efd_val,
+                 thread, t, sysreturn, rv)
+{
+    u64 *efd_val = bound(efd_val);
+    deallocate(bound(h), efd_val, sizeof(*efd_val));
+    closure_finish();
+}
+
+closure_function(4, 2, void, aio_complete,
+                 struct aio *, aio, u64, data, u64, obj, int, res_fd,
+                 thread, t, sysreturn, rv)
+{
+    struct aio *aio = bound(aio);
+    int res_fd = bound(res_fd);
+    aio_ring ring = aio->ring;
+    aio_lock(aio);
+    aio->ongoing_ops--;
+    unsigned int tail = ring->tail;
+    if (tail >= aio->nr) {
```

Comment:
The aio struct is accessible to userspace, so we can't trust it contains valid values. So the apparently redundant checks are there to prevent a process messing with the struct data from crashing the kernel.